8349361: C2: RShiftL should support all applicable transformations that RShiftI does

rwestrel opened the pull request:
This change refactors `RShiftI`/`RshiftL` `Ideal`, `Identity` and
`Value` because the `int` and `long` versions are very similar and so
there's no logic duplication. In the process, support for some extra
transformations is added to `RShiftL`. I also added some new test
cases.

/cc hotspot-compiler

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8349361](https://bugs.openjdk.org/browse/JDK-8349361): C2: RShiftL should support all applicable transformations that RShiftI does (**Enhancement** - P4)


### Reviewers
 * [Emanuel Peter](https://openjdk.org/census#epeter) (@eme64 - **Reviewer**)
 * [Jasmine Karthikeyan](https://openjdk.org/census#jkarthikeyan) (@jaskarth - Committer) Review applies to [e4053783](https://git.openjdk.org/jdk/pull/23438/files/e4053783171e98a2503a3ba274c7bc57be8e493e)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23438/head:pull/23438` \
`$ git checkout pull/23438`

Update a local copy of the PR: \
`$ git checkout pull/23438` \
`$ git pull https://git.openjdk.org/jdk.git pull/23438/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23438`

View PR using the GUI difftool: \
`$ git pr show -t 23438`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23438.diff">https://git.openjdk.org/jdk/pull/23438.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23438#issuecomment-2633851378)
</details>

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,158 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
-//=============================================================================
-//------------------------------Identity---------------------------------------
-Node* RShiftINode::Identity(PhaseGVN* phase) {
+Node *RShiftNode::IdealIL(PhaseGVN *phase, bool can_reshape, BasicType bt) {
```
Drive-by: fix position of `*`

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,158 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+Node *RShiftNode::IdealIL(PhaseGVN *phase, bool can_reshape, BasicType bt) {
+  // Inputs may be TOP if they are dead.
+  const TypeInteger* t1 = phase->type(in(1))->isa_integer(bt);
+  if (!t1) return NodeSentinel;        // Left input is an integer
```
Drive-by: don't use implicit null-check, make comparison with `nullptr` explicit. And add curly braces.

eme64 reviewed (changes requested):
Drive-by code style comment ;)

TobiHartmann commented:
Fails to build on Mac AArch64:

```
[2025-02-05T06:43:04,925Z] * For target hotspot_variant-server_libjvm_objs_mulnode.o:
[2025-02-05T06:43:04,925Z] [...]workspace/open/src/hotspot/share/opto/mulnode.cpp:1400:13: error: use of bitwise '&' with boolean operands [-Werror,-Wbitwise-instead-of-logical]
[2025-02-05T06:43:04,925Z]      assert((checked_cast<jint>(lo) == lo_verify) & (checked_cast<jint>(hi) == hi_verify), "inconsistent");
```

rwestrel commented:
> Fails to build on Mac AArch64:
> 
> ```
> [2025-02-05T06:43:04,925Z] * For target hotspot_variant-server_libjvm_objs_mulnode.o:
> [2025-02-05T06:43:04,925Z] [...]workspace/open/src/hotspot/share/opto/mulnode.cpp:1400:13: error: use of bitwise '&' with boolean operands [-Werror,-Wbitwise-instead-of-logical]
> [2025-02-05T06:43:04,925Z]      assert((checked_cast<jint>(lo) == lo_verify) & (checked_cast<jint>(hi) == hi_verify), "inconsistent");
> ```

Thanks for the report. Should be fixed now. I also took @eme64's comments into account.

jaskarth commented on src/hotspot/share/opto/type.cpp:
```diff
@@ -1530,6 +1530,10 @@ const TypeInteger* TypeInteger::make(jlong lo, jlong hi, int w, BasicType bt) {
   return TypeLong::make(lo, hi, w);
 }
 
+const TypeInteger *TypeInteger::make(jlong lo, BasicType bt) {
```
```suggestion
const TypeInteger* TypeInteger::make(jlong lo, BasicType bt) {
```

jaskarth commented on src/hotspot/share/utilities/globalDefinitions.hpp:
```diff
@@ -792,6 +792,13 @@ inline jlong min_signed_integer(BasicType bt) {
+  if (bt == T_INT) {
+    return BitsPerJavaInteger;
+  }
+  return BitsPerJavaLong;
```
I think it'd be nice to add `assert(bt == T_LONG, "unsupported");` before the last return, like in the helper methods above.

rwestrel replied:
Indeed. Added in new commit.

jaskarth commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1391,136 +1504,26 @@ Node *RShiftINode::Ideal(PhaseGVN *phase, bool can_reshape) {
-    }
-
-    return TypeLong::LONG;
+Node *RShiftLNode::Ideal(PhaseGVN *phase, bool can_reshape) {
```
```suggestion
Node* RShiftLNode::Ideal(PhaseGVN *phase, bool can_reshape) {
```

jaskarth commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+  if (progress != nullptr) {
+    return progress;
   }
+  const TypeInt *t3;  // type of in(1).in(2)
```
```suggestion
  const TypeInt* t3;  // type of in(1).in(2)
```

jaskarth commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
 }
 
+const Type* RShiftNode::ValueIL(PhaseGVN* phase, BasicType bt) const {
+  const Type *t1 = phase->type(in(1));
+  const Type *t2 = phase->type(in(2));
```
```suggestion
  const Type* t1 = phase->type(in(1));
  const Type* t2 = phase->type(in(2));
```

jaskarth commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+
+  // Check for (x & 0xFF000000) >> 24, whose mask can be made smaller.
+  // Such expressions arise normally from shift chains like (byte)(x >> 24).
+  const Node *mask = in(1);
```
```suggestion
  const Node* mask = in(1);
```

jaskarth commented on src/hotspot/share/opto/mulnode.hpp:
```diff
@@ -300,28 +301,42 @@ class RotateRightNode : public TypeNode {
   virtual Node* Identity(PhaseGVN* phase);
-  virtual Node *Ideal(PhaseGVN *phase, bool can_reshape);
+
+  virtual Node* Ideal(PhaseGVN *phase, bool can_reshape);
```
```suggestion
  virtual Node* Ideal(PhaseGVN* phase, bool can_reshape);
```

jaskarth commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+    jlong hi = r1->hi_as_long() >> (jint)shift;
+    assert(lo <= hi, "must have valid bounds");
+#ifdef ASSERT
+   if (bt ==T_INT) {
```
```suggestion
   if (bt == T_INT) {
```
Could this assert be generic to also handle T_LONG too?

rwestrel replied:
The assert checks that, for the int case:

```
long lo;
assert((int)(lo >> shift) == (((int)lo) >> shift, "");
```
For long, it would be:
```
long lo;
assert((long)(lo >> shift) == (((long)lo) >> shift, "");
```
Given everything is already a long, that's:
```
long lo;
assert(lo >> shift == lo >> shift, "");
```

jaskarth replied:
Ah I see, thank you for the explanation!

jaskarth reviewed (commented):
This is really nice! I'd wondered why there was no `RShiftL::Ideal`, and it's nice to have it handled it in a generic way with the integer version. I left mostly code style comments here.

rwestrel commented:
> This is really nice! I'd wondered why there was no `RShiftL::Ideal`, and it's nice to have it handled it in a generic way with the integer version. I left mostly code style comments here.

Thanks for reviewing this. I pushed a new commit that takes your comments into account.

jaskarth reviewed (approved):
Thanks for the update, it looks good!

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
-//=============================================================================
-//------------------------------Identity---------------------------------------
-Node* RShiftINode::Identity(PhaseGVN* phase) {
+Node *RShiftNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
```
```suggestion
Node* RShiftNode::IdealIL(PhaseGVN* phase, bool can_reshape, BasicType bt) {
```

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+  if (t1 == nullptr) {
+    return NodeSentinel;        // Left input is an integer
+  }
+  const TypeInteger* t3;  // type of in(1).in(2)
```
I know that you only moved this code, but it looks bad :see_no_evil: 
For one, why is it defined up here already when it is only used 10 lines later?
And why not give it a better name so we don't need the comment?
```suggestion
```

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+      (t3 = phase->type(mask->in(2))->isa_integer(bt)) &&
+      t3->is_con()) {
+    jlong maskbits = t3->get_con_as_long(bt);
+    // Convert to "(x >> shift) & (mask >> shift)"
```
This is a nice comment. It could come as a motivation above. Because it suggests that we can then constant fold the `mask >> shift`, right?

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+
+  // Check for (x & 0xFF000000) >> 24, whose mask can be made smaller.
+  // Such expressions arise normally from shift chains like (byte)(x >> 24).
+  const Node* mask = in(1);
+  if (mask->Opcode() == Op_And(bt) &&
+      (t3 = phase->type(mask->in(2))->isa_integer(bt)) &&
+      t3->is_con()) {
+    jlong maskbits = t3->get_con_as_long(bt);
```
This is also quite bad. It seems `mask` here is `in(1)`, which is not even the mask at all, but `x & <real_mask>`.

I'd suggest to clean it up a little and use better names.

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+    return TypeInteger::bottom(bt);
+  }
+
+  const TypeInteger* r1 = t1->isa_integer(bt); // Handy access
+  const TypeInt* r2 = t2->isa_int(); // Handy access
```
```suggestion
  const TypeInteger* r1 = t1->isa_integer(bt);
  const TypeInt* r2 = t2->isa_int();
```
Let's reduce the noise a little.

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1296,59 +1308,160 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+  if (progress != nullptr) {
+    return progress;
   }
+  const TypeInt* t3;  // type of in(1).in(2)
```
Also refactor the use of `t3` here, please.

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/RShiftLNodeIdealizationTests.java:
```diff
@@ -37,7 +37,7 @@ public static void main(String[] args) {
     }
 
-    @Run(test = { "test1", "test2", "test3", "test4" })
+    @Run(test = { "test1", "test2", "test3", "test4", "test5", "test6", "test7", "test8", "test9" })
```
You should add the bug id above.

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/RShiftLNodeIdealizationTests.java:
```diff
@@ -89,4 +100,42 @@ public long test3(long x, long y) {
+        return (x >> 128);
+    }
+
+    final int test7Shift = 42;
+    final long test7Min = -1L << (64 - test7Shift -1);
+    final long test7Max = ~test7Min;
```
Could we randomize these tests, so that we would get better coverage?

eme64 reviewed (changes requested):
@rwestrel nice work, looks like a good step to unify the code a little!

I left some comments / suggestions.

I'm also wondering about testing. How good do you think test coverage is? Are all cases covered? How about the edge-cases? Could we improve the coverage with randomization somehow?

rwestrel commented:
@eme64 can you take another look?

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1319,231 +1331,225 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+  const TypeInteger* mask_t;
+  const Node* and_node = in(1);
+  if (and_node->Opcode() == Op_And(bt) &&
+      (mask_t = phase->type(and_node->in(2))->isa_integer(bt)) &&
```
Is this an implicit null check?

Style guide:
> Do not use ints or pointers as (implicit) booleans with &&, ||, if, while. Instead, compare explicitly, i.e. if (x != 0) or if (ptr != nullptr), etc.

eme64 replied:
Honestly, why not just split the `if` into 2 ifs.
`const TypeInteger* mask_t;`
That looks a little less than nice as well.

eme64 replied:
Oh it seams to me this whole code was not really changed, it just appears moved in the diff. Maybe you can rearrange the order so the diff looks smaller, and we don't have to argue about the (questionable) code style here ;)

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1319,231 +1331,225 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
-  // Check for "(short[i] <<16)>>16" which simply sign-extends
-  const Node *shl = in(1);
-  if( shl->Opcode() != Op_LShiftI ) return nullptr;
-
-  if( shift == 16 &&
-      (t3 = phase->type(shl->in(2))->isa_int()) &&
-      t3->is_con(16) ) {
-    Node *ld = shl->in(1);
-    if( ld->Opcode() == Op_LoadS ) {
-      // Sign extension is just useless here.  Return a RShiftI of zero instead
-      // returning 'ld' directly.  We cannot return an old Node directly as
-      // that is the job of 'Identity' calls and Identity calls only work on
-      // direct inputs ('ld' is an extra Node removed from 'this').  The
-      // combined optimization requires Identity only return direct inputs.
-      set_req_X(1, ld, phase);
-      set_req_X(2, phase->intcon(0), phase);
-      return this;
-    }
-    else if (can_reshape &&
-             ld->Opcode() == Op_LoadUS &&
-             ld->outcnt() == 1 && ld->unique_out() == shl)
-      // Replace zero-extension-load with sign-extension-load
-      return ld->as_Load()->convert_to_signed_load(*phase);
-  }
-
-  // Check for "(byte[i] <<24)>>24" which simply sign-extends
-  if( shift == 24 &&
-      (t3 = phase->type(shl->in(2))->isa_int()) &&
-      t3->is_con(24) ) {
-    Node *ld = shl->in(1);
-    if (ld->Opcode() == Op_LoadB) {
-      // Sign extension is just useless here
-      set_req_X(1, ld, phase);
-      set_req_X(2, phase->intcon(0), phase);
-      return this;
-    }
-  }
-
-  return nullptr;
-}
```
This is the same code as above, right? As commented above: it would be good to not move it and reduce the size of the diff.

rwestrel replied:
I reordered the changes. Can you take another look?

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/RShiftLNodeIdealizationTests.java:
```diff
@@ -89,4 +101,47 @@ public long test3(long x, long y) {
+    @IR(counts = { IRNode.RSHIFT_L, "1", IRNode.LSHIFT_L, "1" })
+    public long test9(long x) {
+        x = Integer.max(Integer.min((int)x, (int)test7Max), (int)(test7Min-1));
+        return ((x << test7Shift) >> test7Shift);
```
It could be nice to have some test cases where both shift values are completely randomized.

rwestrel replied:
The transformation only happens if the amounts we shift left and right are the same. So if they are random, the transformation won't apply most of the time and, rarely, it will (because they will turn out to be the same). I'm not sure how to write an IR test then.

eme64 replied:
You would not have to assert anything about the IR, just do value verification in that case.

rwestrel replied:
Added in new commit. Is that what you had in mind?

rwestrel commented on test/hotspot/jtreg/compiler/c2/irTests/RShiftLNodeIdealizationTests.java:
```diff
@@ -89,4 +101,47 @@ public long test3(long x, long y) {
+        return (x >> test6Shift);
+    }
+
+    final static int test7Shift = RunInfo.getRandom().nextInt(32) + 32;
+    final static long test7Min = -1L << (64 - test7Shift -1);
+    final static long test7Max = ~test7Min;
```
Added in new commit.

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1321,69 +1333,104 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
-  if( shift == 16 &&
-      (t3 = phase->type(shl->in(2))->isa_int()) &&
-      t3->is_con(16) ) {
+  const TypeInt* left_shift_t;
+  if (shift == 16 &&
+      (left_shift_t = phase->type(shl->in(2))->isa_int()) &&
+      left_shift_t->is_con(16)) {
```
```suggestion
  const TypeInt* left_shift_t = phase->type(shl->in(2))->isa_int();
  if (shift == 16 &&
      left_shift_t != nullptr &&
      left_shift_t->is_con(16)) {
```
Would that be equivalent? I think it would be more readable.

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1401,9 +1448,9 @@ Node *RShiftINode::Ideal(PhaseGVN *phase, bool can_reshape) {
-      (t3 = phase->type(shl->in(2))->isa_int()) &&
-      t3->is_con(24) ) {
+  if (shift == 24 &&
+      (left_shift_t = phase->type(shl->in(2))->isa_int()) &&
```
Then you could also refactor down here.
Because these are implicit nullptr checks again.

eme64 commented on src/hotspot/share/opto/mulnode.hpp:
```diff
@@ -318,28 +319,42 @@ class RotateRightNode : public TypeNode {
+
+class RShiftNode : public Node {
+public:
+  RShiftNode(Node* in1, Node* in2) : Node(nullptr,in1,in2) {}
```
```suggestion
  RShiftNode(Node* in1, Node* in2) : Node(nullptr, in1, in2) {}
```

eme64 commented on src/hotspot/share/opto/mulnode.hpp:
```diff
@@ -318,28 +319,42 @@ class RotateRightNode : public TypeNode {
+class RShiftINode : public RShiftNode {
 public:
-  RShiftINode( Node *in1, Node *in2 ) : Node(nullptr,in1,in2) {}
+  RShiftINode(Node* in1, Node* in2) : RShiftNode(in1,in2) {}
```
```suggestion
  RShiftINode(Node* in1, Node* in2) : RShiftNode(in1, in2) {}
```

eme64 commented on src/hotspot/share/opto/mulnode.hpp:
```diff
@@ -318,28 +319,42 @@ class RotateRightNode : public TypeNode {
+class RShiftLNode : public RShiftNode {
 public:
-  RShiftLNode( Node *in1, Node *in2 ) : Node(nullptr,in1,in2) {}
+  RShiftLNode(Node* in1, Node* in2) : RShiftNode(in1,in2) {}
```
```suggestion
  RShiftLNode(Node* in1, Node* in2) : RShiftNode(in1, in2) {}
```

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/RShiftINodeIdealizationTests.java:
```diff
@@ -37,7 +37,7 @@ public static void main(String[] args) {
     }
 
-    @Run(test = { "test1", "test2", "test3", "test4" })
+    @Run(test = { "test1", "test2", "test3", "test4", "test5", "test6", "test7", "test8", "test9", "test10" })
```
For completeness sake, you should add the bug number above too ;)

eme64 reviewed (approved):
@rwestrel Thanks for moving code so it is easier to review :blush: 

I have a few more minor suggestions, but on the whole I'm happy with it :)

I'd still like to run another round of testing once we have it all finished up though, so please ping me again once you worked through my comments ;)

rwestrel commented:
> I'd still like to run another round of testing once we have it all finished up though, so please ping me again once you worked through my comments ;)

@eme64 new commit should address all your comments.

eme64 commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1321,69 +1333,106 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
+  return nullptr;
+}
+
+Node *RShiftINode::Ideal(PhaseGVN *phase, bool can_reshape) {
```
```suggestion
Node* RShiftINode::Ideal(PhaseGVN* phase, bool can_reshape) {
```

eme64 commented on test/hotspot/jtreg/compiler/c2/irTests/RShiftLNodeIdealizationTests.java:
```diff
@@ -64,6 +68,15 @@ public void assertResult(long x, long y) {
+        Asserts.assertEQ((x8 << test7Shift) >> test7Shift, test8(x));
+        long x9 = Integer.max(Integer.min((int)x, (int)test7Max), (int)(test7Min-1));
+        Asserts.assertEQ((x9 << test7Shift) >> test7Shift, test9(x));
+         Asserts.assertEQ(((x7 << test7Shift) >> test10Shift), test10(x));
```
```suggestion
        Asserts.assertEQ(((x7 << test7Shift) >> test10Shift), test10(x));
```

eme64 reviewed (changes requested):
Did another quick scan. Will launch some testing now.

chhagedorn commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1321,69 +1333,106 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
-      int hi = ~lo;               // 00007FFF
-      const TypeInt* t11 = phase->type(in(1)->in(1))->isa_int();
+      jlong lo = (-1 << (bits_per_java_integer(bt) - ((uint)count)-1)); // FFFF8000
+      jlong hi = ~lo;               // 00007FFF
```
Seems strangely align. Maybe either align it to the comment above or convert to an unaligned comment.

chhagedorn commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1321,69 +1333,106 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
     }
     // Check for useless sign-masking
-    if (in(1)->Opcode() == Op_LShiftI &&
+    if (in(1)->Opcode() == Op_LShift(bt) &&
         in(1)->req() == 3 &&
         in(1)->in(2) == in(2)) {
```
Generally, is there notifaction code for this pattern to re-add the node to the IGVN worklist? If not, I don't think you need to handle it here if it's missing (it's just a missed opportunity but no correctness issue) but would be good to file a follow-up bug to handle it - especially when we want to add IGVN verification for `Ideal` and `Identity` with [JDK-8347273](https://bugs.openjdk.org/browse/JDK-8347273).

rwestrel replied:
Good catch. I fixed a case that was handled for int but not for long. There are others that are missing for both AFAICT.
If I file a follow up bug, writing a test case for it is going to be very hard. So testing a fix is also hard. Shouldn't we wait for JDK-8347273 and fix whatever follows up of that?

chhagedorn replied:
> If I file a follow up bug, writing a test case for it is going to be very hard

Yes, it will be. There is currently no way to really verify that reliably without the additional verification. I think it's okay to wait for JDK-8347273. Maybe you can add a note there or file a separate issue to keep track of the missing bits detected in this investigation..

chhagedorn commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1321,69 +1333,106 @@ const Type* LShiftLNode::Value(PhaseGVN* phase) const {
-      t3->is_con() ) {
-    Node *x = mask->in(1);
-    jint maskbits = t3->get_con();
+  const Node* and_node = in(1);
```
Same as above, do we have notification code for this patterns checking? Same for the patterns in `RShiftINode::Ideal()`.

chhagedorn commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1416,136 +1463,95 @@ Node *RShiftINode::Ideal(PhaseGVN *phase, bool can_reshape) {
     uint shift = r2->get_con();
-    shift &= BitsPerJavaInteger-1;  // semantics of Java shifts
-    // Shift by a multiple of 32 does nothing:
+    shift &= bits_per_java_integer(bt)-1;  // semantics of Java shifts
```
```suggestion
    shift &= bits_per_java_integer(bt) - 1;  // semantics of Java shifts
```

chhagedorn commented on src/hotspot/share/opto/mulnode.cpp:
```diff
@@ -1416,136 +1463,95 @@ Node *RShiftINode::Ideal(PhaseGVN *phase, bool can_reshape) {
-    if (shift == BitsPerJavaInteger-1) {
-      if (r1->_lo >= 0) assert(ti == TypeInt::ZERO,    ">>31 of + is  0");
-      if (r1->_hi <  0) assert(ti == TypeInt::MINUS_1, ">>31 of - is -1");
+    if (shift == bits_per_java_integer(bt)-1) {
```
```suggestion
    if (shift == bits_per_java_integer(bt) - 1) {
```

chhagedorn commented on src/hotspot/share/opto/mulnode.hpp:
```diff
@@ -318,28 +319,42 @@ class RotateRightNode : public TypeNode {
 
+
+class RShiftNode : public Node {
+public:
```
```suggestion
 public:
```

chhagedorn commented on src/hotspot/share/opto/type.cpp:
```diff
@@ -1699,6 +1699,10 @@ const TypeInteger* TypeInteger::make(jlong lo, jlong hi, int w, BasicType bt) {
   return TypeLong::make(lo, hi, w);
 }
 
+const TypeInteger* TypeInteger::make(jlong lo, BasicType bt) {
```
Maybe you want to rename `lo` to `con` since we set `lo == hi`.

chhagedorn commented on test/hotspot/jtreg/compiler/c2/irTests/RShiftINodeIdealizationTests.java:
```diff
@@ -27,7 +27,7 @@
 
```
Up to you if you want to update the copyright year or add your company's copyright. Same in the other test.

chhagedorn reviewed (approved):
Nice refactoring! I have a few small comments - mostly code style. Otherwise, looks good to me, too.

rwestrel commented:
> Nice refactoring! I have a few small comments - mostly code style. Otherwise, looks good to me, too.

Thanks for the review.
New commit should address all your comments.
Now that the long min/max intrinsic is integrated, I also changed the long tests so they use long min/max and that triggered a bug in the code (missing `CONST64`) that I fixed.

eme64 commented:
@rwestrel I saw this in testing from yesterday: `linux-aarch64-debug` with `-XX:+UnlockExperimentalVMOptions -XX:+UseCompactObjectHeaders`.

```
Failed IR Rules (2) of Methods (2)
----------------------------------
1) Method "public long compiler.c2.irTests.RShiftLNodeIdealizationTests.test8(long)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#RSHIFT_L#_", "1", "_#LSHIFT_L#_", "1"}, failOn={}, applyIfPlatform={}, applyIfPlatformOr={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(RShiftL.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
         * Constraint 2: "(\\d+(\\s){2}(LShiftL.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!

2) Method "public long compiler.c2.irTests.RShiftLNodeIdealizationTests.test9(long)" - [Failed IR rules: 1]:
   * @IR rule 1: "@compiler.lib.ir_framework.IR(phase={DEFAULT}, applyIfPlatformAnd={}, applyIfCPUFeatureOr={}, counts={"_#RSHIFT_L#_", "1", "_#LSHIFT_L#_", "1"}, failOn={}, applyIfPlatform={}, applyIfPlatformOr={}, applyIfOr={}, applyIfCPUFeatureAnd={}, applyIf={}, applyIfCPUFeature={}, applyIfAnd={}, applyIfNot={})"
     > Phase "PrintIdeal":
       - counts: Graph contains wrong number of nodes:
         * Constraint 1: "(\\d+(\\s){2}(RShiftL.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
         * Constraint 2: "(\\d+(\\s){2}(LShiftL.*)+(\\s){2}===.*)"
           - Failed comparison: [found] 0 = 1 [given]
           - No nodes matched!
```

eme64 commented:
Not sure if that still reproduces after your changes. LMK when I should run testing again.

chhagedorn reviewed (approved):
Apart from Emanuel's report, it looks good to me, thanks for the update!

rwestrel commented:
> Not sure if that still reproduces after your changes. LMK when I should run testing again.

I could reproduce that one by forcing one of the random value to a particular constant. It's a bug in the test that was fix since. @eme64 could you run testing again, please?

rwestrel commented:
@eme64 any update on testing?

eme64 commented:
@rwestrel Testing looks good, thanks for the ping :)

eme64 reviewed (approved):
Gave it a quick pass again, I think this is good to go (though better to integrate after the weekend...) 

@rwestrel thanks for the work and all the updates :relaxed:

rwestrel commented:
@eme64 thanks for running tests and for the review.
@chhagedorn @jaskarth thanks for the reviews.

rwestrel commented:
/integrate
